filter.go

```diff
@@ -8,60 +8,150 @@ import (
 	}
 }
 
+func (f *Filters) find(ff *net.IPNet) int {
+	ffs := ff.String()
+	for idx, ft := range f.filters {
+		if ft.f.String() == ffs {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-	fs.filters[f.String()] = f
+
+	idx := fs.find(f)
+	if idx != -1 {
+		fs.filters[idx].reject = true
+	} else {
+		fs.filters = append(fs.filters, &filterEntry{f: f, reject: true})
+	}
+}
+
+// AddAllowFilter adds an accept rule to the given Filters. Hosts
+// matching the given net.IPNet filter will be accepted, unless
+// another policy is added which states that they should be rejected.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
+func (fs *Filters) AddAllowFilter(f *net.IPNet) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+
+	idx := fs.find(f)
+	if idx != -1 {
+		fs.filters[idx].reject = false
+	} else {
+		fs.filters = append(fs.filters, &filterEntry{f: f, reject: false})
+	}
+}
+
+// Remove removes all net.IPNet's accept/reject rule(s) from the
+// Filters, if there are matching rules.
+//
+// Makes no distinction between whether the rule is an allow or a
+// deny.
+func (f *Filters) Remove(ff *net.IPNet) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	idx := f.find(ff)
+	if idx != -1 {
+		f.filters = append(f.filters[:idx], f.filters[idx+1:]...)
+	}
 }
 
+// AddrBlocked parses a ma.Multiaddr and, if it can get a valid netip
+// back, applies the Filters returning true if the given address
+// should be rejected, and false if the given address is allowed.
+//
+// If a parsing error occurs, or no filter matches, the Filters
+// default is returned.
 func (f *Filters) AddrBlocked(a ma.Multiaddr) bool {
 	maddr := ma.Split(a)
 	if len(maddr) == 0 {
-		return false
+		return f.RejectByDefault
 	}
 	netaddr, err := manet.ToNetAddr(maddr[0])
 	if err != nil {
-		// if we can't parse it, it's probably not blocked
-		return false
+		// if we cant parse it, its probably not blocked
+		return f.RejectByDefault
 	}
 	netip := net.ParseIP(netaddr.String())
 	if netip == nil {
-		return false
+		return f.RejectByDefault
 	}
 
 	f.mu.RLock()
 	defer f.mu.RUnlock()
+
+	var reject bool = f.RejectByDefault
+
 	for _, ft := range f.filters {
-		if ft.Contains(netip) {
-			return true
+		if ft.f.Contains(netip) {
+			reject = ft.reject
 		}
 	}
-	return false
+
+	return reject
 }
 
+// Filters returns the list of DENY net.IPNet masks
 func (f *Filters) Filters() []*net.IPNet {
 	var out []*net.IPNet
 	f.mu.RLock()
 	defer f.mu.RUnlock()
 	for _, ff := range f.filters {
-		out = append(out, ff)
+		if ff.reject {
+			out = append(out, ff.f)
+		}
```

Comment:
and we should probably call this `DenyFilters`.... but how is `RejectByDefault` handled here semantically?